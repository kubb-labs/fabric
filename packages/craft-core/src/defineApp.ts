import { write } from './fs.ts'
import { createFile, parseFile } from './parser.ts'
import type { Ref } from './reactive/ref.ts'
import type * as KubbFile from './types.ts'

const isFunction = (val: unknown): val is Function => typeof val === 'function'

type Component = any

type PluginInstallFunction<Options = any[]> = Options extends unknown[] ? (app: App, ...options: Options) => any : (app: App, options: Options) => any

export type ObjectPlugin<Options = any[]> = {
  install: PluginInstallFunction<Options>
}
export type FunctionPlugin<Options = any[]> = PluginInstallFunction<Options> & Partial<ObjectPlugin<Options>>

type Renderer = {
  run(): Promise<void> | void
  files: Ref<Array<KubbFile.File>>
  waitUntilExit(): Promise<void>
  output: Ref<string>
}

export type RootRenderFunction<HostElement = unknown> = (container: HostElement) => Renderer

export type Plugin<Options = any[], P extends unknown[] = Options extends unknown[] ? Options : [Options]> = FunctionPlugin<P> | ObjectPlugin<P>

export interface App<_HostElement = unknown> {
  _component: Component
  run(): Promise<void>
  use<Options>(plugin: Plugin<Options>, options: NoInfer<Options>): this
  write(): Promise<void>
  files: Array<KubbFile.File>
  waitUntilExit(): Promise<void>
  output: string
}

export type DefineApp<HostElement> = (rootComponent: Component) => App<HostElement>

export function defineApp<HostElement>(instance: RootRenderFunction<HostElement>): DefineApp<HostElement> {
  function createApp<HostElement>(rootComponent: Component): App<HostElement> {
    const installedPlugins = new WeakSet()
    const { run, waitUntilExit, files, output } = instance(rootComponent)

    const app: App<HostElement> = {
      _component: rootComponent,
      get files() {
        return files.value
      },
      get output() {
        return output.value
      },
      async run() {
        await run()

        await waitUntilExit()
      },
      waitUntilExit,
      async write() {
        for (const file of files.value) {
          const resolvedFile = createFile(file)
          const source = await parseFile(resolvedFile)

          await write(file.path, source)
        }
      },
      use(plugin: Plugin, ...options: any[]) {
        if (installedPlugins.has(plugin)) {
          console.warn('Plugin has already been applied to target app.')
        } else if (plugin && isFunction(plugin.install)) {
          installedPlugins.add(plugin)
          plugin.install(app, ...options)
        } else if (isFunction(plugin)) {
          installedPlugins.add(plugin)
          plugin(app, ...options)
        }

        return app
      },
    }

    return app
  }

  return createApp
}
