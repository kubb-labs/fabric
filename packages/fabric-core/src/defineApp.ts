import { FileManager } from './FileManager.ts'
import { isFunction } from 'remeda'
import type { Plugin } from './plugins/types.ts'
import type { Parser } from './parsers/types.ts'
import { AsyncEventEmitter } from './utils/AsyncEventEmitter.ts'
import type { AppContext, Component, AppEvents } from './App.ts'

import type { App } from './index.ts'

type RootRenderFunction<TApp extends App> = (app: TApp) => void | Promise<void>

export type DefineApp<TOptions> = (rootComponent?: Component, options?: TOptions) => App

export function defineApp<TOptions = unknown>(instance?: RootRenderFunction<App<TOptions>>): DefineApp<TOptions> {
  function createApp(options?: TOptions): App {
    const events = new AsyncEventEmitter<AppEvents>()
    const installedPlugins = new Set<Plugin>()
    const installedParsers = new Set<Parser>()
    const fileManager = new FileManager({ events })
    const context = {
      events,
      options,
      fileManager,
      installedPlugins,
      installedParsers,
    } as AppContext<TOptions>

    const app = {
      context,
      get files() {
        return fileManager.files
      },
      async addFile(...newFiles) {
        await fileManager.add(...newFiles)
      },
      async use(pluginOrParser, ...options) {
        const args = options

        if (pluginOrParser.type === 'plugin') {
          if (installedPlugins.has(pluginOrParser)) {
            console.warn('Plugin has already been applied to target app.')
          } else {
            installedPlugins.add(pluginOrParser)
          }

          if (pluginOrParser.inject && isFunction(pluginOrParser.inject)) {
            const injecter = pluginOrParser.inject

            const extraApp = (injecter as any)(app, ...args)
            Object.assign(app, extraApp)
          }
        }
        if (pluginOrParser.type === 'parser') {
          if (installedParsers.has(pluginOrParser)) {
            console.warn('Parser has already been applied to target app.')
          } else {
            installedParsers.add(pluginOrParser)
          }
        }

        if (pluginOrParser && isFunction(pluginOrParser.install)) {
          const installer = pluginOrParser.install

          await (installer as any)(app, ...args)
        }

        return app
      },
    } as App<TOptions>

    if (instance) {
      instance(app)
    }

    return app
  }

  return createApp
}
