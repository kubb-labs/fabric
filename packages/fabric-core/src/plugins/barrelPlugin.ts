/** biome-ignore-all lint/suspicious/useIterableCallbackReturn: not needed */

import { createPlugin } from './createPlugin.ts'
import type * as KubbFile from '../KubbFile.ts'
import { TreeNode } from '../utils/TreeNode.ts'
import path from 'node:path'
import { getRelativePath } from '../utils/getRelativePath.ts'
import { createFile } from '../createFile.ts'

type Mode = 'all' | 'named' | 'propagate' | false

type Options = {
  root: string
  mode: Mode
  dryRun?: boolean
}

type WriteEntryOptions = {
  root: string
  mode: Mode
}

type ExtendOptions = {
  writeEntry(options: WriteEntryOptions): Promise<void>
}

// biome-ignore lint/suspicious/noTsIgnore: production ready
// @ts-ignore
declare module '@kubb/fabric-core' {
  interface App {
    writeEntry(options: WriteEntryOptions): Promise<void>
  }
}

declare global {
  namespace Kubb {
    interface App {
      writeEntry(options: WriteEntryOptions): Promise<void>
    }
  }
}

type GetBarrelFilesOptions = {
  files: KubbFile.File[]
  root: string
  mode: Mode
}

export function getBarrelFiles({ files, root, mode }: GetBarrelFilesOptions): Array<KubbFile.File> {
  // Do not generate when propagating or disabled
  if (mode === 'propagate' || mode === false) {
    return []
  }

  const cachedFiles = new Map<KubbFile.Path, KubbFile.File>()
  const dedupe = new Map<KubbFile.Path, Set<string>>()

  const treeNode = TreeNode.fromFiles(files, root)

  if (!treeNode) {
    return []
  }

  treeNode.forEach((node) => {
    // Only create a barrel for directory-like nodes that have a parent with a path
    if (!node?.children || !node.parent?.data.path) {
      return
    }

    const parentPath = node.parent.data.path as KubbFile.Path
    const barrelPath = path.join(parentPath, 'index.ts') as KubbFile.Path

    let barrelFile = cachedFiles.get(barrelPath)
    if (!barrelFile) {
      barrelFile = createFile({
        path: barrelPath,
        baseName: 'index.ts',
        exports: [],
        sources: [],
      })
      cachedFiles.set(barrelPath, barrelFile)
      dedupe.set(barrelPath, new Set<string>())
    }

    const seen = dedupe.get(barrelPath)!

    // Collect all leaves under the current directory node
    node.leaves.forEach((leaf) => {
      const file = leaf.data.file
      if (!file) {
        return
      }

      const sources = file.sources || []
      sources.forEach((source) => {
        if (!file.path || !source.isIndexable || !source.name) {
          return
        }

        const key = `${source.name}|${source.isTypeOnly ? '1' : '0'}`
        if (seen.has(key)) {
          return
        }
        seen.add(key)

        // Always compute relative path from the parent directory to the file path
        barrelFile!.exports!.push({
          name: [source.name],
          path: getRelativePath(parentPath, file.path),
          isTypeOnly: source.isTypeOnly,
        })

        barrelFile!.sources.push({
          name: source.name,
          isTypeOnly: source.isTypeOnly,
          value: '', // TODO use parser to generate import
          isExportable: mode === 'all' || mode === 'named',
          isIndexable: mode === 'all' || mode === 'named',
        })
      })
    })
  })

  const result = [...cachedFiles.values()]

  if (mode === 'all') {
    return result.map((file) => ({
      ...file,
      exports: file.exports?.map((e) => ({ ...e, name: undefined })),
    }))
  }

  return result
}

export const barrelPlugin = createPlugin<Options, ExtendOptions>({
  name: 'barrel',
  install(app, options) {
    if (!options) {
      throw new Error('Barrel plugin requires options.root and options.mode')
    }

    if (!options.mode) {
      return undefined
    }

    app.context.events.on('write:start', async ({ files }) => {
      const root = options.root
      const barrelFiles = getBarrelFiles({ files, root, mode: options.mode })

      await app.context.fileManager.add(...barrelFiles)
    })
  },
  inject(app, options) {
    if (!options) {
      throw new Error('Barrel plugin requires options.root and options.mode')
    }

    return {
      async writeEntry({ root, mode }) {
        if (!mode || mode === 'propagate') {
          return undefined
        }

        const rootPath = path.resolve(root, 'index.ts')

        const barrelFiles = app.files.filter((file) => {
          return file.sources.some((source) => source.isIndexable)
        })

        const entryFile = createFile({
          path: rootPath,
          baseName: 'index.ts',
          exports: barrelFiles
            .flatMap((file) => {
              const containsOnlyTypes = file.sources.every((source) => source.isTypeOnly)

              return file.sources
                ?.map((source) => {
                  if (!file.path || !source.isIndexable) {
                    return undefined
                  }

                  return {
                    name: mode === 'all' ? undefined : [source.name],
                    path: getRelativePath(rootPath, file.path),
                    isTypeOnly: mode === 'all' ? containsOnlyTypes : source.isTypeOnly,
                  } as KubbFile.Export
                })
                .filter(Boolean)
            })
            .filter(Boolean),
          sources: [],
        })

        await app.context.fileManager.add(entryFile)

        await app.context.fileManager.write({
          mode: app.context.options?.mode,
          dryRun: options.dryRun,
          parsers: app.context.installedParsers,
        })
      },
    }
  },
})
