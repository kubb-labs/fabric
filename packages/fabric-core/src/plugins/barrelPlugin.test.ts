import { beforeEach, describe, expect, test, vi } from 'vitest'

import { barrelPlugin, getBarrelFiles } from './barrelPlugin.ts'
import { createFile } from '../createFile.ts'
import { defineApp } from '../defineApp.ts'

const files = [
  createFile({
    path: 'src/a.ts',
    baseName: 'a.ts',
    sources: [{ name: 'A', isExportable: true, isIndexable: true }],
  }),
  createFile({
    path: 'src/sub/sub2/b.ts',
    baseName: 'b.ts',
    sources: [{ name: 'b', isExportable: true, isIndexable: true }],
  }),
  createFile({
    path: 'src/sub/sub2/c.ts',
    baseName: 'c.ts',
    sources: [{ name: 'C', isExportable: false, isIndexable: false }],
  }),
  createFile({
    path: 'src/sub/d.ts',
    baseName: 'd.ts',
    sources: [
      { name: 'D', isExportable: true, isIndexable: true },
      { name: 'E', isExportable: true, isIndexable: true, isTypeOnly: true },
    ],
  }),
  createFile({
    path: 'src/sub/index.ts',
    baseName: 'index.ts',
    sources: [
      {
        name: 'world',
        value: 'export const world = 2;',
        isExportable: true,
        isIndexable: true,
      },
      {
        name: 'hello',
        value: 'export const hello = 2;',
        isExportable: true,
        isIndexable: true,
      },
    ],
  }),
]

describe('getBarrelFiles', () => {
  beforeEach(() => {
    vi.restoreAllMocks()
  })
  test("mode 'all' should produce wildcard exports and mark barrel sources indexable/exportable", () => {
    const barrelFiles = getBarrelFiles({ files, root: 'src', mode: 'all' })

    expect(barrelFiles).toMatchSnapshot()
    expect(barrelFiles.length).toBe(3)

    expect(barrelFiles.every((file) => file.baseName === 'index.ts')).toBeTruthy()
    expect(barrelFiles.every((f) => f.sources.every((s) => s.isIndexable && s.isExportable))).toBe(true)

    const rootIndex = barrelFiles.find((f) => f.path === 'src/index.ts')!
    expect(rootIndex.exports?.length).toBe(6)
  })

  test("mode 'named' should produce named exports and mark barrel sources indexable/exportable", () => {
    const barrelFiles = getBarrelFiles({ files, root: 'src', mode: 'named' })

    expect(barrelFiles).toMatchSnapshot()
    expect(barrelFiles.length).toBe(3)

    expect(barrelFiles.every((file) => file.baseName === 'index.ts')).toBeTruthy()
    expect(barrelFiles.every((f) => f.sources.every((s) => s.isIndexable && s.isExportable))).toBe(true)

    const rootIndex = barrelFiles.find((f) => f.path === 'src/index.ts')!
    expect(rootIndex.exports?.length).toBe(6)
  })

  test("mode 'propagate' should not generate any barrel files", () => {
    const barrelFiles = getBarrelFiles({ files, root: 'src', mode: 'propagate' })
    expect(barrelFiles.length).toBe(0)
  })

  test('mode false should not generate any barrel files', () => {
    const barrelFiles = getBarrelFiles({ files, root: 'src', mode: false })
    expect(barrelFiles.length).toBe(0)
  })
})

describe('barrelPlugin', () => {
  const example = {
    baseName: 'index.ts',
    exports: [
      {
        isTypeOnly: false,
        name: undefined,
        path: './a.ts',
      },
      {
        isTypeOnly: false,
        name: undefined,
        path: './sub/d.ts',
      },
      {
        isTypeOnly: false,
        name: undefined,
        path: './sub/d.ts',
      },
      {
        isTypeOnly: false,
        name: undefined,
        path: './sub/index.ts',
      },
      {
        isTypeOnly: false,
        name: undefined,
        path: './sub/index.ts',
      },
      {
        isTypeOnly: false,
        name: undefined,
        path: './sub/sub2/b.ts',
      },
    ],
  }
  const exampleNamed = {
    baseName: 'index.ts',
    exports: [
      {
        isTypeOnly: undefined,
        name: ['A'],
        path: './a.ts',
      },
      {
        isTypeOnly: undefined,
        name: ['D'],
        path: './sub/d.ts',
      },
      {
        isTypeOnly: true,
        name: ['E'],
        path: './sub/d.ts',
      },
      {
        isTypeOnly: undefined,
        name: ['world'],
        path: './sub/index.ts',
      },
      {
        isTypeOnly: undefined,
        name: ['hello'],
        path: './sub/index.ts',
      },
      {
        isTypeOnly: undefined,
        name: ['b'],
        path: './sub/sub2/b.ts',
      },
    ],
    path: '/Users/stijnvanhulle/GitHub/fabric/src/index.ts',
    sources: [],
  }

  beforeEach(() => {
    vi.restoreAllMocks()
  })
  test("mode 'all' should produce wildcard exports and mark barrel sources indexable/exportable", () => {
    const app = defineApp()()

    app.use(barrelPlugin, { mode: 'propagate', root: 'src' })
    app.addFile(...files)

    const addSpy = vi.spyOn(app.context.fileManager, 'add')

    app.writeEntry({
      root: 'src',
      mode: 'all',
    })

    expect(addSpy).toHaveBeenCalledTimes(1)
    expect(addSpy).toHaveBeenCalledWith(expect.objectContaining(example))
  })

  test("mode 'named' should produce named exports and mark barrel sources indexable/exportable", async () => {
    const app = defineApp()()

    app.use(barrelPlugin, { mode: 'propagate', root: 'src' })
    await app.addFile(...files)

    const addSpy = vi.spyOn(app.context.fileManager, 'add')

    await app.writeEntry({
      root: 'src',
      mode: 'named',
    })

    expect(addSpy).toHaveBeenCalledTimes(1)
    expect(addSpy).toHaveBeenCalledWith(expect.objectContaining(exampleNamed))
  })

  test("mode 'propagate' should not generate any barrel files", () => {
    const app = defineApp()()

    app.use(barrelPlugin, { mode: 'propagate', root: 'src' })
    app.addFile(...files)

    const addSpy = vi.spyOn(app.context.fileManager, 'add')

    app.writeEntry({
      root: 'src',
      mode: 'propagate',
    })

    expect(addSpy).toHaveBeenCalledTimes(0)
  })
  test('mode false should not generate any barrel files', async () => {
    const app = defineApp()()

    app.use(barrelPlugin, { mode: 'propagate', root: 'src' })
    await app.addFile(...files)

    const addSpy = vi.spyOn(app.context.fileManager, 'add')

    await app.writeEntry({
      root: 'src',
      mode: false,
    })

    expect(addSpy).toHaveBeenCalledTimes(0)
  })
})
