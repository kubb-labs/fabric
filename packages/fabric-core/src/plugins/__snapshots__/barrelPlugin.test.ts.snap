// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`barrelPlugin > mode 'all' should produce wildcard exports and mark barrel sources indexable/exportable 1`] = `
[
  {
    "isTypeOnly": false,
    "name": undefined,
    "path": "./a.ts",
  },
  {
    "isTypeOnly": false,
    "name": undefined,
    "path": "./sub/d.ts",
  },
  {
    "isTypeOnly": false,
    "name": undefined,
    "path": "./sub/index.ts",
  },
  {
    "isTypeOnly": false,
    "name": undefined,
    "path": "./sub/sub2/b.ts",
  },
]
`;

exports[`barrelPlugin > mode 'named' should produce named exports and mark barrel sources indexable/exportable 1`] = `
[
  {
    "isTypeOnly": true,
    "name": [
      "E",
    ],
    "path": "./sub/d.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "A",
    ],
    "path": "./a.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "D",
    ],
    "path": "./sub/d.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "world",
      "hello",
    ],
    "path": "./sub/index.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "b",
    ],
    "path": "./sub/sub2/b.ts",
  },
]
`;

exports[`getBarrelFiles > mode 'all' should produce wildcard exports and mark barrel sources indexable/exportable 1`] = `
[
  {
    "isTypeOnly": undefined,
    "name": undefined,
    "path": "./a.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": undefined,
    "path": "./sub/d.ts",
  },
  {
    "isTypeOnly": true,
    "name": undefined,
    "path": "./sub/d.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": undefined,
    "path": "./sub/index.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": undefined,
    "path": "./sub/index.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": undefined,
    "path": "./sub/sub2/b.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": undefined,
    "path": "./sub2/b.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": undefined,
    "path": "./d.ts",
  },
  {
    "isTypeOnly": true,
    "name": undefined,
    "path": "./d.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": undefined,
    "path": "./index.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": undefined,
    "path": "./index.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": undefined,
    "path": "./b.ts",
  },
]
`;

exports[`getBarrelFiles > mode 'named' should produce named exports and mark barrel sources indexable/exportable 1`] = `
[
  {
    "isTypeOnly": undefined,
    "name": [
      "A",
    ],
    "path": "./a.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "D",
    ],
    "path": "./sub/d.ts",
  },
  {
    "isTypeOnly": true,
    "name": [
      "E",
    ],
    "path": "./sub/d.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "world",
    ],
    "path": "./sub/index.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "hello",
    ],
    "path": "./sub/index.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "b",
    ],
    "path": "./sub/sub2/b.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "b",
    ],
    "path": "./sub2/b.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "D",
    ],
    "path": "./d.ts",
  },
  {
    "isTypeOnly": true,
    "name": [
      "E",
    ],
    "path": "./d.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "world",
    ],
    "path": "./index.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "hello",
    ],
    "path": "./index.ts",
  },
  {
    "isTypeOnly": undefined,
    "name": [
      "b",
    ],
    "path": "./b.ts",
  },
]
`;
